File: src/main/java/FastSlowPointer/Q3.java
package FastSlowPointer;
//Middle of the LinkedList
public class Q3 {

    public static  ListNode2 findMiddle(ListNode2 head) {
        ListNode2 slow = head;
        ListNode2 fast = head;

        while(fast != null && fast.next != null && fast.next.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        return slow;
    }

    public static void main(String[] args) {
        ListNode2 head = new ListNode2(1);
        head.next = new ListNode2(2);
        head.next.next = new ListNode2(3);
        head.next.next.next = new ListNode2(4);
        head.next.next.next.next = new ListNode2(5);

        System.out.println("head: "+head.val);

        ListNode2 middle = findMiddle(head);
        System.out.println("middle: "+middle.val);
    }

    public static class ListNode2 {
        int val;
        ListNode2 next;
        ListNode2(int x) {
            this.val = x;
            this.next = null;
        }

    }
}

File: scala/FastSlowPointer/Q3.scala
//Problem Challenge 1 - Palindrome LinkedList

class ListNode(var _x: Int = 0) {
    var next: ListNode = null
}

def isPalindrome(head: ListNode): Boolean = {

    def reverseList(node: ListNode): ListNode = {
        var prev: ListNode = null
        var curr = node

        while(curr != null) {
            val nextTemp = curr.next
            curr.next=  prev
            prev = curr
            curr = nextTemp
        }
        prev
    }
 
   
    var slow = head
    var fast = head

    while(fast != null && fast.next != null) {
        slow = slow.next
        fast = fast.next.next
    }
    if(fast != null) slow = slow.next

    var p1 = head
    var p2 = reverseList(slow)

    while(p2 != null) {
        if(p1._x != p2._x) return false
        p1 = p1.next
        p2 = p2.next
    }
    true
}

def main(args: Array[String]): Unit = {
    
    val head = new ListNode(1)
    val node1 = new ListNode(2)
    val node2 = new ListNode(2)
    val node3 = new ListNode(1)

    head.next = node1
    node1.next = node2
    node2.next = node3

    println("Is Palindrome: "+isPalindrome(head))

}
